File: src/security.ts
import * as path from "path";

/**
 * Sanitize a slug to prevent path traversal and injection.
 * Allows alphanumeric, hyphens, and underscores only.
 */
export function sanitizeSlug(slug: string): string {
  return slug
    .toLowerCase()
    .replace(/[^a-z0-9-_]/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
}

/**
 * Resolve a file path within the project root, rejecting traversal attempts.
 */
export function safePath(projectRoot: string, ...segments: string[]): string {
  const resolved = path.resolve(projectRoot, ...segments);
  const normalized = path.normalize(resolved);
  if (!normalized.startsWith(path.resolve(projectRoot))) {
    throw new Error(`Path traversal attempt detected: ${segments.join("/")}`);
  }
  return normalized;
}

/**
 * Zero-pad a number to a fixed width.
 */
export function padId(n: number, width = 4): string {
  return String(n).padStart(width, "0");
}

File: research_md/security.py
"""Path traversal prevention and slug sanitization."""

import os
import re


def safe_path(root: str, *parts: str) -> str:
    resolved = os.path.normpath(os.path.join(os.path.abspath(root), *parts))
    if not resolved.startswith(os.path.abspath(root)):
        raise ValueError(f"Path traversal attempt detected: {'/'.join(parts)}")
    return resolved


def sanitize_slug(slug: str) -> str:
    s = slug.lower()
    s = re.sub(r"[^a-z0-9\-_]", "-", s)
    s = re.sub(r"-+", "-", s)
    s = s.strip("-")
    return s


def pad_id(n: int, width: int = 4) -> str:
    return str(n).zfill(width)
